src/templating.js

```diff
@@ -1,308 +1,349 @@
+        default:
+            throw new Error("Unknown renderMode: " + renderMode);
     }
 
-    function getFirstNodeFromPossibleArray(nodeOrNodeArray) {
-        return nodeOrNodeArray.nodeType ? nodeOrNodeArray
-                                        : nodeOrNodeArray.length > 0 ? nodeOrNodeArray[0]
-                                        : null;
+    if (haveAddedNodesToParent) {
+        activateBindingsOnContinuousNodeArray(renderedNodesArray, bindingContext);
+        if (options['afterRender'])
+            dependencyDetection.ignore(options['afterRender'], null, [renderedNodesArray, bindingContext['$data']]);
     }
 
-    function executeTemplate(targetNodeOrNodeArray, renderMode, template, bindingContext, options) {
-        options = options || {};
-        var firstTargetNode = targetNodeOrNodeArray && getFirstNodeFromPossibleArray(targetNodeOrNodeArray);
-        var templateDocument = (firstTargetNode || template || {}).ownerDocument;
-        var templateEngineToUse = (options['templateEngine'] || _templateEngine);
-        ko.templateRewriting.ensureTemplateIsRewritten(template, templateEngineToUse, templateDocument);
-        var renderedNodesArray = templateEngineToUse['renderTemplate'](template, bindingContext, options, templateDocument);
-
-        // Loosely check result is an array of DOM nodes
-        if ((typeof renderedNodesArray.length != "number") || (renderedNodesArray.length > 0 && typeof renderedNodesArray[0].nodeType != "number"))
-            throw new Error("Template engine must return an array of DOM nodes");
-
-        var haveAddedNodesToParent = false;
-        switch (renderMode) {
-            case "replaceChildren":
-                ko.virtualElements.setDomNodeChildren(targetNodeOrNodeArray, renderedNodesArray);
-                haveAddedNodesToParent = true;
-                break;
-            case "replaceNode":
-                ko.utils.replaceDomNodes(targetNodeOrNodeArray, renderedNodesArray);
-                haveAddedNodesToParent = true;
-                break;
-            case "ignoreTargetNode": break;
-            default:
-                throw new Error("Unknown renderMode: " + renderMode);
-        }
+    return renderedNodesArray;
+}
+
+function resolveTemplateName(template, data, context) {
+    // The template can be specified as:
+    if (isObservable(template)) {
+        // 1. An observable, with string value
+        return template();
+    } else if (typeof template === 'function') {
+        // 2. A function of (data, context) returning a string
+        return template(data, context);
+    } else {
+        // 3. A string
+        return template;
+    }
+}
+
+function keyValueArrayContainsKey(keyValueArray, key) {
+    for (var i = 0; i < keyValueArray.length; i++)
+        if (keyValueArray[i]['key'] == key)
+            return true;
+    return false;
+}
+
+export function renderTemplate(template, dataOrBindingContext, options, targetNodeOrNodeArray, renderMode) {
```

Comment:
This *feels* somewhat overly complex, and that may just me being naive, but I feel like if we drop IE<8 support this gets simpler ... but again ... probably me being naive. :grin: